### ISSUE TITLE
Apache Commons bean fields not generated

### ISSUE THREAD

**Issue by jdussouillez** [vojtechhabarta/typescript-generator#290]
I'm trying to generate the typescript code for a bean that contains an [org.apache.commons.lang3.tuple.ImmutablePair](https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/tuple/ImmutablePair.java) object but the fields are not present in the generated TS file.

**Edit:**
When I implement a custom class which extends Pair<L,R> the problem is the same.

---

Bean :
```java
package com.test.bean;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.net.HostAndPort;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.tuple.ImmutablePair;

public class MyBean {

    @Getter
    @Setter
    @JsonProperty
    protected String s;

    @Getter
    @Setter
    @JsonProperty
    protected ImmutablePair<String, Boolean> pair;

    @Getter
    @Setter
    @JsonProperty
    protected HostAndPort host;
}
```

Maven configuration : 
```xml
<plugin>
    <groupId>cz.habarta.typescript-generator</groupId>
    <artifactId>typescript-generator-maven-plugin</artifactId>
    <version>2.8.449</version>
    <executions>
        <execution>
            <id>generate</id>
            <phase>process-classes</phase>
            <goals>
                <goal>generate</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <jsonLibrary>jackson2</jsonLibrary>
        <classPatterns>
            <classPattern>com.test.bean.**</classPattern>
        </classPatterns>
        <excludeClasses>
            <excludeClass>java.io.Serializable</excludeClass>
            <excludeClass>java.lang.Comparable</excludeClass>
        </excludeClasses>
        <mapClasses>asClasses</mapClasses>
        <mapDate>asDate</mapDate>
        <mapEnum>asEnum</mapEnum>
        <nonConstEnums>true</nonConstEnums>
        <mapPackagesToNamespaces>true</mapPackagesToNamespaces>
        <optionalProperties>useLibraryDefinition</optionalProperties>
        <outputFile>model.ts</outputFile>
        <outputFileType>implementationFile</outputFileType>
        <outputKind>module</outputKind>
    </configuration>
</plugin>
```
Generated code : 

```typescript
/* tslint:disable */
// Generated using typescript-generator version 2.8.449 on 2018-11-09 13:59:01.

export namespace com.test.bean {

    export class MyBean {
        s?: string;
        pair?: org.apache.commons.lang3.tuple.ImmutablePair<string, boolean>;
        host?: com.google.common.net.HostAndPort;
    }

}

export namespace org.apache.commons.lang3.tuple {

    export class Pair<L, R> implements java.util.Map.Entry<L, R> {
    }

}

export namespace org.apache.commons.lang3.tuple {

    export class ImmutablePair<L, R> extends org.apache.commons.lang3.tuple.Pair<L, R> {
        // ------------------------------
        // NOTHING HERE ?!
        // ------------------------------
    }

}

export namespace com.google.common.net {

    export class HostAndPort {
        host?: string;
        port?: number;
    }

}

export namespace java.util.Map {

    export interface Entry<K, V> {
    }

}
```

Environment : 
```
$ java -version
openjdk version "11" 2018-09-25
OpenJDK Runtime Environment (build 11+24-Ubuntu-116.04.1)
OpenJDK 64-Bit Server VM (build 11+24-Ubuntu-116.04.1, mixed mode, sharing)
```

```
$ mvn -version
Apache Maven 3.3.9
Maven home: /usr/share/maven
Java version: 11, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.10.0-38-generic", arch: "amd64", family: "unix"
```

**Comment by vojtechhabarta** [comment]
I tried to serialize instance of `Pair` to JSON using Jackson and the result is following:

``` json
{
  "pair": {
    "left": "right"
  }
}
```
which is object with single property where the key is "stringified" left part of the pair and the value is right part of the pair.

I also realized that `Pair<L, R>` implements `Map.Entry<L, R>`. So I think we could add general support for `Map.Entry<K, V>`. It is not probably possible to generate TypeScript type which would exactly describe these objects. I think closest type would be `{ [index: string]: V }`. It defines object with string keys and values of type `V` but it doesn't limit number of properties in this object to one.

Or do you have better idea what should be generated?

**Comment by jdussouillez** [comment]
The Jackson serialization of `Map.Entry` used to serialize it as an object key/value [but it changed a few years ago](https://github.com/fasterxml/jackson-databind/issues/565#issuecomment-78129048).
I agree that in this case, the `{ [index: string]: V }` syntax is the most appropriated one.

However, in my real case, the key of my Pair is an object. So I need the `Map.Entry<K, V>` class to be generated as a TS Object `{key: K, value: V}`. Maybe we can set an option to change the generation method of Map.Entry classes ?

As you can see [here](https://github.com/fasterxml/jackson-databind/issues/565#issuecomment-317821695), you can force the serialization to make Jackson generate the key/value fields in an object (see the test file below). Do you think we could have an option to enable the same behavior ?


```java
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.AbstractMap;
import java.util.Map;
import static org.junit.jupiter.api.Assertions.assertEquals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class JacksonTest {

    protected ObjectMapper mapper;
    protected Map.Entry<String, Object> entry;
    protected Map.Entry<MyBean, Object> entryBeanKey;

    @BeforeEach
    public void setUp() {
        mapper = new ObjectMapper();
        entry = new AbstractMap.SimpleEntry<>("foo", 0);
        entryBeanKey = new AbstractMap.SimpleEntry<>(new MyBean("bar"), 1);
    }

    @Test
    public void testJacksonSerialize() throws JsonProcessingException {
        assertEquals("{\"foo\":0}", mapper.writeValueAsString(entry));
        assertEquals("{\"MyBean instance\":1}", mapper.writeValueAsString(entryBeanKey));
    }

    @Test
    public void testJacksonSerializeObject() throws JsonProcessingException {
        mapper.configOverride(Map.Entry.class)
            .setFormat(JsonFormat.Value.forShape(JsonFormat.Shape.OBJECT));

        assertEquals("{\"key\":\"foo\",\"value\":0}", mapper.writeValueAsString(entry));
        assertEquals("{\"key\":{\"f0\":\"bar\",\"f1\":true},\"value\":1}", mapper.writeValueAsString(entryBeanKey));
    }

    private static class MyBean {

        @JsonProperty
        protected final String f0;

        @JsonProperty
        protected boolean f1;

        public MyBean(final String f0) {
            this.f0 = f0;
            f1 = true;
        }

        @Override
        public String toString() {
            return "MyBean instance";
        }
    }
}
```

**Comment by vojtechhabarta** [comment]
Ok, I will investigate it more, namely if we can get this info directly from Jackson library.
Thanks for examples.

**Comment by vojtechhabarta** [comment]
I implemented support for types derived from `Map.Entry` which are serialized by Jackson into two different formats (shapes). Here is recapitulation how Jackson serializes `Map.Entry` and what typescript-generator generates in these cases.

For `JsonFormat.Shape.NATURAL` Jackson produces JSON object with single property with **string** key:
``` json
{
  "someKey": "someValue"
}
```
while for `JsonFormat.Shape.OBJECT` Jackson produces JSON object with `key` and `value` properties:
``` json
{
  "key": "someKey",
  "value": "someValue"
}
```

For these cases following types are generated:
``` typescript
interface ClassWithEntries {
    entry1: { [index: string]: SomeValue };  // NATURAL shape
    entry2: Entry<SomeKey, SomeValue>;  // OBJECT shape
}

interface Entry<K, V> {
    key: K;
    value: V;
}
``` 

Jackson has two ways how to specify which shape should be used: annotations and configuration overrides:
``` java
// annotation
@JsonFormat(shape = JsonFormat.Shape.OBJECT)
class MyEntry<K, V> extends AbstractMap.SimpleEntry<K, V> {}

// override
objectMapper.configOverride(MyEntry.class)
    .setFormat(JsonFormat.Value.forShape(JsonFormat.Shape.OBJECT));
```

Both ways are now supported in typescript-generator. Here is sample Maven configuration with format overrides:
``` xml
<configuration>
    <jsonLibrary>jackson2</jsonLibrary>
    <jackson2Configuration>
        <shapeConfigOverrides>
            <override>java.util.Map$Entry:OBJECT</override>
            <override>com.package.MyEntry:OBJECT</override>
        </shapeConfigOverrides>
    </jackson2Configuration>
    ...
</configuration>
```
Gradle snippet:
``` gradle
generateTypeScript {
    jsonLibrary = 'jackson2'
    jackson2Configuration = [
        shapeConfigOverrides: [
            'java.util.Map$Entry:OBJECT',
            'com.package.MyEntry:OBJECT',
        ]
    ]
}
```

**Comment by vojtechhabarta** [comment]
@jdussouillez could you please review it and check if it works as expected?

**Comment by jdussouillez** [comment]
@vojtechhabarta Thank you ! It's exactly what I need.

---

Note: Just be careful with this example : because `Pair` provides `getLeft`/`getRight` and `Map.Entry` provides `getKey`/`getValue`, the generated TS is :

```typescript
export namespace org.apache.commons.lang3.tuple {

    export class Pair<L, R> {
        value?: R;
        key?: L;
        right?: R;
        left?: L;
    }
}
```

But this can be fixed using a subclass and the Jackson `@JsonIgnore` annotation on the undesired getters. Otherwise, everything works great !

**Comment by jdussouillez** [comment]
Do you have a date for the next release including this new feature ? I'm using the [maven repository](https://mvnrepository.com/artifact/cz.habarta.typescript-generator/typescript-generator-maven-plugin) for my project but the recent tags are not available yet.

Thanks

**Comment by vojtechhabarta** [comment]
Released [v2.10.466](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.10.466).
